### DISCUSSION TITLE
Unusual P-F Runtime Error

### DISCUSSION THREAD

**Discussion by jsevic** [idaholab/moose#19088] [Q&A Modules: General] [answered]
Hello,

I have been working with the P-F for some time now, and recently this run time error came up. I replaced the TimeDerivative operator with HeatConductionTimeDerivative operator to pass thermal properties and the attached screen capture show the run  time error. I set the kernel back to TimeDerivative and still get the error.

It seems like the make file is not creating new code for the linked HeatConductionTimeDerivative operator, at least the dates on the linked files are not changing.

Any help would be appreciated thx!

John Sevic

![Screen Shot 2021-10-12 at 7 21 50 PM](https://user-images.githubusercontent.com/4711942/137056561-07145f84-5611-4ec3-9c78-51efedf184f0.png)

**Comment by GiudGiud** [comment, marked as answer]
Hello

This type of error happens when some linked libraries are recompiled and the executable cannot find routines anymore. 
Possibly here, you recompiled the framework but did not recompile the phase field module
It just means you need to recompile. 
You may have to `make clobberall` first.

Guillaume

**Reply by jsevic** [reply]
Guillaume,

OK thanks I ran the clobber command and the compiler started on all the code and the EulerAngleProvider (and some other Euler functions) have been returning this error? Any suggestions?

I have been using the P-F app I made in 2019 without any issues and then I added some new kernels and these errors showed up, this is the back ground.

Thank you.

John Sevic
![Screen Shot 2021-10-13 at 7 41 10 AM](https://user-images.githubusercontent.com/4711942/137156541-a5f77a11-d111-4d51-8461-d86eb0accc28.png)

**Reply by jsevic** [reply]
Please also see attached another screen capture...
![Screen Shot 2021-10-13 at 7 45 19 AM](https://user-images.githubusercontent.com/4711942/137156882-3867dc0c-4925-4766-a177-c2a90235788e.png)

**Reply by GiudGiud** [reply]
Did you update moose and libmesh?

**Comment by jsevic** [comment]
I have not, since I all my code has been working fine up to now?

**Reply by GiudGiud** [reply]
Well it s always good to update. You ll get new stuff and less bugs. 

But anyway, 
It's not finding Matrix and EulerAngles.
I'm guessing you are missing a header, like #include "EulerAngles.h". It's in tensor_mechanics utils.

**Reply by GiudGiud** [reply]
Ah actually you have EulerAngles, but it s looking for Matrix. Lemme find where that is

**Reply by GiudGiud** [reply]
well the content of EulerAngles.h has changed it seems
```
//* This file is part of the MOOSE framework
//* https://www.mooseframework.org
//*
//* All rights reserved, see COPYRIGHT for full restrictions
//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
//*
//* Licensed under LGPL 2.1, please see LICENSE for details
//* https://www.gnu.org/licenses/lgpl-2.1.html

#pragma once

#include "MooseTypes.h"
#include "libmesh/vector_value.h"
// Ignore depcrecated copy warnings for this class from libmesh Eigen contrib
#include "libmesh/ignore_warnings.h"
#include <Eigen/Geometry>

// forward declaration
class MooseRandom;

/**
 * Euler angle triplet.
 */
class EulerAngles
{
public:
  Real phi1, Phi, phi2;

  // default constructor
  EulerAngles();
  // Quaternions to Euler Angles
  EulerAngles(Eigen::Quaternion<Real> & q);

  operator RealVectorValue() const { return RealVectorValue(phi1, Phi, phi2); }

  void random();
  void random(MooseRandom & random);
  // Euler to Quaternions
  Eigen::Quaternion<Real> toQuaternion();
};

// Restore warnings for other classes
#include "libmesh/restore_warnings.h"

```
so I ll need your moose commit number if you really dont want to update

**Comment by jsevic** [comment]
Ok cool let me see if I can get it compile exploring some of these Euler angle files. I can update on another machine. Thx!

**Reply by jsevic** [reply]
Sorry, how do I  get my MOOSE commit number, was this from my initial install?

**Reply by jsevic** [reply]
And I will do a fresh install tonight too, just in case.

**Reply by GiudGiud** [reply]
it's the last commit that is not from your work. 
You can see the log of commits with `git log`

**Reply by jsevic** [reply]
OK I will have to dig that up, the last install I did was many years ago.

Attempting to install the latest MOOSE produced the following...
![Screen Shot 2021-10-14 at 6 59 44 AM](https://user-images.githubusercontent.com/4711942/137333023-e8e15036-5c13-4e8a-b9a1-8f334846687e.png)

**Reply by GiudGiud** [reply]
We prefer you paste test directly (in triple quotes for readability) rather than post screenshots

It seems your conda is struggling from the beginning. After installing conda, did you follow the prompts it gave you? It should ask for you to
`conda init $your shell` then restart the terminal

**Comment by jsevic** [comment]
Ok so you want me to the terminal text and put it in quotes no problem.

Yes it dis restart my terminal. FYI I am running El Cap.

**Reply by jsevic** [reply]
Guillaume,

Before moving further on this, I did want to point out what is puzzling - I wrote the present app in 2019 and have made various changes to the physics, all of this of course requiring compiling, and I have never seen this error before. In fact, compiling generally does not produce any errors for the present app, except this one time, from some files that don't seem to have any obvious connection to P-F? I may try reverting back to a known good build and work from the direction to see where it falls apart.

Thx!

John Sevic

**Reply by GiudGiud** [reply]
Are you using the conda build or did you install your own libmesh?

**Comment by jsevic** [comment]
I installed libmesh - I don’t recall conda being around when I did the first install in 2016.

**Reply by jsevic** [reply]
OK that is interesting my back-up drive failed so reversion is presently not an option. I attempted a new install of the current MOOSE release and this is what I see.

Note that I am currently running 10.11.6 (El Cap) and that the conda install did not finish correctly, does this need to install correctly in order for MOOSE to compile?



John-Sevics-MBP-4:~ jsevic$ cd projects
John-Sevics-MBP-4:projects jsevic$ git clone https://github.com/idaholab/moose.git
Cloning into 'moose'...
remote: Enumerating objects: 511520, done.
remote: Total 511520 (delta 0), reused 0 (delta 0), pack-reused 511520
Receiving objects: 100% (511520/511520), 401.27 MiB | 12.86 MiB/s, done.
Resolving deltas: 100% (389042/389042), done.
Checking out files: 100% (29514/29514), done.
John-Sevics-MBP-4:projects jsevic$ cd moose
John-Sevics-MBP-4:moose jsevic$ git checkout master
Branch 'master' set up to track remote branch 'master' from 'origin'.
Switched to a new branch 'master'
John-Sevics-MBP-4:moose jsevic$ cd ~/projects/moose/test
John-Sevics-MBP-4:test jsevic$ make -j 4
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/libtool: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/libtool: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/jsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
env: python3: No such file or directory
/Users/jsevic/projects/moose/framework/moose.mk:15: *** Build failed.  Stop.
John-Sevics-MBP-4:test jsevic$

**Reply by jsevic** [reply]
And just in case, this is  the error returned on the conda install...

Last login: Sun Oct 17 09:48:07 on ttys000
John-Sevics-MBP-4:~ jsevic$ export PATH=$HOME/miniconda3/bin:$PATH
John-Sevics-MBP-4:~ jsevic$ conda config --add channels conda-forge
Warning: 'conda-forge' already in 'channels' list, moving to the top
John-Sevics-MBP-4:~ jsevic$ conda config --add channels https://conda.software.inl.gov/public
Warning: 'https://conda.software.inl.gov/public' already in 'channels' list, moving to the top
John-Sevics-MBP-4:~ jsevic$ conda create --name moose -q -y
Collecting package metadata (current_repodata.json): ...working... done
Solving environment: ...working... failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): ...working... done
Solving environment: ...working... failed

PackagesNotFoundError: The following packages are not available from current channels:

  - moose

Current channels:

  - https://conda.software.inl.gov/public/osx-64
  - https://conda.software.inl.gov/public/noarch
  - https://conda.anaconda.org/conda-forge/osx-64
  - https://conda.anaconda.org/conda-forge/noarch
  - https://repo.anaconda.com/pkgs/main/osx-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/osx-64
  - https://repo.anaconda.com/pkgs/r/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.


John-Sevics-MBP-4:~ jsevic$

**Reply by GiudGiud** [reply]
@milljm do we have conda packages for el capitan?

conda installation does need to succeed to install moose after.

Otherwise can you just run scripts/update_and_rebuild_libmesh ? After having git fetch then git pull latest moose
@milljm do we have a page for manual installation - updating? my quick search did not succeed in finding it

**Comment by jsevic** [comment]
Ok let me try. Thx!

**Reply by jsevic** [reply]
Any guidance on MOOSE installing without Conda? Does the current MOOSE support El Cap still?

Thx!

**Reply by GiudGiud** [reply]
What is your xcode version ?
/usr/bin/xcodebuild -version

you need at least Xcode SDK 10.10 for MOOSE

**Comment by jsevic** [comment]
I am running v. 9.

Presently I believe the most efficient path forward is to get a new Mac Mini, for several reasons, so let me install the newest MOOSE on this new machine, it has the latest release of iOS and associated tools.

I will get back to this  thread if there are any issues.

Thx!

**Reply by GiudGiud** [reply]
Are you talking about the M1 minis? You'll still have to be mindful about the XCode version, but the speed boost will definitely be worth it for compiling
See https://github.com/idaholab/moose/discussions/18134

**Comment by jsevic** [comment]
Ah that is a good question, for the current MOOSE release, what version of XC is required/supported?

Yeah the M1 and I get more RAM too ✔️

Thx!

**Reply by loganharbour** [reply]
FYI - the M1 is not officially supported. We’re making progress on it, but it’s not complete.

**Comment by jsevic** [comment]
Cool, in fact I got an Intel Mini just in case so thanks for telling me.

**Reply by jsevic** [reply]
OK so my new Mac Mini has arrived, it is an Intel Late 2018 machine with Big Sur. Can I go directly to the current MOOSE install that appears to be set up for Catalina? Is X-Code 13 the correct version to support the current release of MOOSE?

Thx!
John Sevic

**Reply by GiudGiud** [reply]
XCode 13 (and M1 possibly) will be supported for sure when this https://github.com/idaholab/moose/pull/19230 gets merged. 
In the mean time we currently support 12.4 on Big Sur. 

Guillaume

**Comment by jsevic** [comment]
Ok cool I just loaded Monterey and X-Code 13 I will try loading MOOSE and conda next thx!

**Comment by jsevic** [comment]
OK so at the moment I have upgraded to Monterey (12.01) and X-Code 13 and I am getting a ton of errors run when the compile for tests.

**Comment by jsevic** [comment]
(base) johnsevic@Johns-Mac-mini test % make -j 4
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/libtool: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/libtool: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
Using HIT from /Users/johnsevic/projects/moose/framework/contrib/hit
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
Compiling C++ (in opt mode) /Users/johnsevic/projects/moose/framework/build/unity_src/src_Unity.C...
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
Compiling C++ (in opt mode) /Users/johnsevic/projects/moose/framework/build/unity_src/linesearches_Unity.C...
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/libtool: No such file or directory
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/contrib/bin/libmesh-config: No such file or directory
make: *** [/Users/johnsevic/projects/moose/framework/build/unity_src/src_Unity..opt.lo] Error 127
make: *** Waiting for unfinished jobs....
/bin/sh: /Users/johnsevic/projects/moose/libmesh/installed/libtool: No such file or directory
make: *** [/Users/johnsevic/projects/moose/framework/build/unity_src/linesearches_Unity..opt.lo] Error 127

**Reply by GiudGiud** [reply]
you forgot to activate the conda environment

**Comment by jsevic** [comment]
Also my libmesh directory is empty, it does not look like anything landed in there on the install?

**Reply by GiudGiud** [reply]
libmesh is installed by conda somewhere else. moose/libmesh is only populated for the manual install

**Comment by GiudGiud** [comment]
@jsevic may I suggest you use these commands to do this:

```
cd ~/projects
git clone https://github.com/idaholab/moose.git moose_test
 
conda deactivate
conda install mamba 
conda create -y -n moose-libmesh-test
conda activate moose-libmesh-test
mamba install -c https://conda.software.inl.gov/moose/pr-19230 moose-libmesh moose-tools 

# Open new terminal session
conda activate moose-libmesh-test
cd ~/projects/moose_test/test
make -j 8
./run_tests -j 8
```

@cticenhour and @loganharbour are working on getting the support working for the latest OS and Xcode. I don't think it will work unless you use this.

**Reply by GiudGiud** [reply]
as you can see from all the test suffixes, this is something we are testing internally right now. Please report to us if it does not work we would like to hear about it. 

And once this PR is merged, you will have to change the mamba channel.

**Comment by jsevic** [comment]
OK so yes this gets me to an earlier issue, when I did activate MOOSE the compiler starts to run but produces hundreds of lines of errors, should I paste them all or just a snippet?

**Reply by GiudGiud** [reply]
Yeah that s because the channel you pulled libmesh from isn't working right now with the latest Xcode. See my post above.

**Comment by jsevic** [comment]
OK so do I need to revert back to XC 12?

**Reply by cticenhour** [reply]
If you look in your message history, @GiudGiud gave you the instructions of how to set up a working MOOSE environment for your system. Follow those instructions to get a working installation while we finish up the work to get it officially supported.

**Reply by GiudGiud** [reply]
Either that or try to follow the instructions I posted. I think the latter is easier.

**Comment by jsevic** [comment]
Oh wait i missed your comment on mamba let me read it...

**Comment by jsevic** [comment]
Sweet it's compiling thx!

**Comment by jsevic** [comment]
All tests pass thx!

**Comment by jsevic** [comment]
@GiudGiud I think you asked me to report how this is working and I have compiled several examples, some of my own apps, and the P-F module has compiled but I have not run any tests or apps yet. It looks like some new functions have been added to P-F.
Thx!
John

**Reply by GiudGiud** [reply]
glad to hear it. Thanks!

**Comment by jsevic** [comment]
@GiudGiud so I tried adding some additional kernels from the heat conduction module and it looks like the kernel registration process has changed? I used to register the kernel in the my_app.c file in /src/base but the new file shows something quite different, did this process change?

**Reply by GiudGiud** [reply]
Yeah i dont know about this old registration process. 
We register them at the top of their own source file now, see the register call:
https://github.com/idaholab/moose/blob/next/framework/src/kernels/ADDiffusion.C

**Comment by jsevic** [comment]
OK so it sounds like on adding a custom kernel, say my_kernel.c, in that file I would have
registerMooseObject("MyApp", my_kernel); and in this file there would also be a #include "my_kernel.h" and these two files would be in their usual locations?
Thx!

**Reply by GiudGiud** [reply]
yup

**Comment by jsevic** [comment]
OK that is cool and I understand that, now it seems I have a compile issue again. I followed the instructions, restarted a couple of times, and from my command line I activate conda and moose lib-mesh and moose and I get the the folloiwing.

I am sure this cockpit error since I compiled this app already, thanks!

John




(base) johnsevic@x86_64-apple-darwin13 heat_conduction % conda activate
(base) johnsevic@x86_64-apple-darwin13 heat_conduction % conda activate moose-libmesh-test                
(moose-libmesh-test) johnsevic@x86_64-apple-darwin13 heat_conduction % conda activate moose             
(moose) johnsevic@x86_64-apple-darwin13 heat_conduction % make -j 8                                        
Using HIT from /Users/johnsevic/projects/moose/moose_test/framework/contrib/hit
Compiling C++ (in opt mode) /Users/johnsevic/projects/moose/moose_test/modules/heat_conduction/build/unity_src/kernels_Unity.C...
In file included from /Users/johnsevic/projects/moose/moose_test/modules/heat_conduction/build/unity_src/kernels_Unity.C:2:
In file included from /Users/johnsevic/projects/moose/moose_test/modules/heat_conduction/src/kernels/ADHeatConduction.C:10:
In file included from /Users/johnsevic/projects/moose/moose_test/modules/heat_conduction/build/header_symlinks/ADHeatConduction.h:12:
In file included from /Users/johnsevic/projects/moose/moose_test/framework/build/header_symlinks/ADDiffusion.h:12:
In file included from /Users/johnsevic/projects/moose/moose_test/framework/build/header_symlinks/ADKernelGrad.h:12:
In file included from /Users/johnsevic/projects/moose/moose_test/framework/build/header_symlinks/ADKernel.h:12:
In file included from /Users/johnsevic/projects/moose/moose_test/framework/build/header_symlinks/KernelBase.h:12:
In file included from /Users/johnsevic/projects/moose/moose_test/framework/build/header_symlinks/ResidualObject.h:12:
In file included from /Users/johnsevic/projects/moose/moose_test/framework/build/header_symlinks/MooseObject.h:13:
In file included from /Users/johnsevic/projects/moose/moose_test/framework/build/header_symlinks/InputParameters.h:13:
In file included from /Users/johnsevic/projects/moose/moose_test/framework/build/header_symlinks/MooseUtils.h:13:
In file included from /Users/johnsevic/projects/moose/moose_test/framework/build/header_symlinks/HashMap.h:13:
In file included from /Users/johnsevic/miniconda3/envs/moose/libmesh/include/libmesh/threads.h:24:
In file included from /Users/johnsevic/miniconda3/envs/moose/libmesh/include/libmesh/libmesh_common.h:36:
In file included from /Users/johnsevic/miniconda3/envs/moose/bin/../include/c++/v1/cstdlib:85:
/Users/johnsevic/miniconda3/envs/moose/bin/../include/c++/v1/stdlib.h:142:34: error: unknown type name 'ldiv_t'
inline _LIBCPP_INLINE_VISIBILITY ldiv_t div(long __x, long __y) _NOEXCEPT {
                                 ^
/Users/johnsevic/miniconda3/envs/moose/bin/../include/c++/v1/stdlib.h:143:12: error: no member named 'ldiv' in the global namespace
  return ::ldiv(__x, __y);
         ~~^

**Reply by GiudGiud** [reply]
hello

You can only activate one moose environment, and it needs to be up-to-date. 

I think the last one you activated `conda activate moose`, is not up to date. 
You can update it with `conda update --all` once it's activated. 

Guillaume

**Comment by jsevic** [comment]
OK I ran these commands and moose updated however I am getting the same compile error, did I miss something? Recall I was told to use this moose version - git clone https://github.com/idaholab/moose.git moose_test

thank you!
john

**Reply by GiudGiud** [reply]
that was a temporary fix because you were on Monterey and we did not support that at the time. Please make sure to use the vanilla instructions from now on. 

Can you please try following these instructions?
https://mooseframework.inl.gov/getting_started/installation/update_moose.html

**Comment by jsevic** [comment]
Ok to be clear then is it correct Monterey is now supported via a standard moose release? Thx.

**Reply by GiudGiud** [reply]
yes! except for M1 which is still WIP but you're on an intel I think.
And peacock may still have issues

**Comment by jsevic** [comment]
OK cool  thanks I will follow the instructions you just posted.

**Reply by jsevic** [reply]
I am getting the same type of errors, should I delete the initial moose directory and start over from scratch? thx!

**Comment by jsevic** [comment]
I went ahead and tried a new install and when I got the conda install lib-mesh I got the following...

(moose) johnsevic@x86_64-apple-darwin13 ~ % conda install lib-mesh   
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.

PackagesNotFoundError: The following packages are not available from current channels:

  - lib-mesh

Current channels:

  - https://conda.software.inl.gov/public/osx-64
  - https://conda.software.inl.gov/public/noarch
  - https://conda.anaconda.org/conda-forge/osx-64
  - https://conda.anaconda.org/conda-forge/noarch
  - https://repo.anaconda.com/pkgs/main/osx-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/osx-64
  - https://repo.anaconda.com/pkgs/r/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.


(moose) johnsevic@x86_64-apple-darwin13 ~ %

**Reply by GiudGiud** [reply]
Try libmesh instead of lib-mesh

**Reply by GiudGiud** [reply]
But I think the package name is actually moose-libmesh. 
there s a page with the exact instructions you can just copy paste

**Comment by jsevic** [comment]
worked thx! let me try running tests next thx!

**Comment by jsevic** [comment]
ok so I ran tests and then also compiled heat_conduction and phase_field and ran tests and they both worked. Next I copied over some kernels from heat_conduction to phase_field (eventually for electrothermal P-F) and I changed the app reference to PhaseFieldApp in each of the .c kernels I copied over and I get this...

(moose) johnsevic@x86_64-apple-darwin13 phase_field % make   
Using HIT from /Users/johnsevic/projects/moose/framework/contrib/hit
Linking Library /Users/johnsevic/projects/moose/modules/phase_field/lib/libphase_field-opt.la...
duplicate symbol 'ElectricalConductivityTempl<false>::setDerivatives(double&, double, MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long long, MetaPhysicL::NWrapper<50ul> >, true> const&)' in:
    /Users/johnsevic/projects/moose/modules/phase_field/build/unity_src/.libs/kernels_Unity.x86_64-apple-darwin13.4.0.opt.o
    /Users/johnsevic/projects/moose/modules/phase_field/build/unity_src/.libs/materials_Unity.x86_64-apple-darwin13.4.0.opt.o
ld: 1 duplicate symbol for architecture x86_64
clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [/Users/johnsevic/projects/moose/framework/app.mk:336: /Users/johnsevic/projects/moose/modules/phase_field/lib/libphase_field-opt.la] Error 1
(moose) johnsevic@x86_64-apple-darwin13 phase_field %

**Reply by GiudGiud** [reply]
I think something is getting compiled twice. 

Did you add header guards in your header? (#pragma once at the top of the header)

**Comment by jsevic** [comment]
did i miss a file somewhere?

**Comment by jsevic** [comment]
I have copied to appropriate directories in phase_field to heat conduction these kernels: HeatConduction, HeatConductionTimeDerivative, HeatSource, and JouleHeatingSource. I also copied the material ElectricalConductivity. They all seem to have the #pragma once in the .h file I am sorry I don't know what this means. In the .c files I did register the app to PhaseFieldApp.

Thx!

**Comment by jsevic** [comment]
I was able to compile by removing the kernel ElectricalConductivity from the src directory, so that seems to suggest the problem lies with this kernel. I placed in the /src/materials directory.

**Comment by jsevic** [comment]
On this kernel there are the following two registrations...
registerMooseObject("PhaseFieldApp", ElectricalConductivity);
registerMooseObject("PhaseFieldApp", ADElectricalConductivity);

The first one makes sense to me, I am not sure what the second is and I don't see and associated .c file anywhere?

**Reply by GiudGiud** [reply]
the second one is for the automatic differentiation version of the kernel. 
ElectricalConductivityTempl is a template for a kernel, that can return regular residuals (ElectricalConductivity) or automatic-differentiation augmented residuals (ADElectricalConductivity). From the latter, you can extract a perfect numerical Jacobian, which is really neat. 

Ok but now I dont understand, why are you moving files from one module to another? 
You can just change the Makefile in your app to include all the modules you need. 

The error with setDerivatives in that template could be removed by deleting one of the versions I guess
It really shouldnt return a compiler error but I m at a loss so it s worth a try

**Comment by jsevic** [comment]
ok so this is something new for me, supposing my app's starting point is PhaseFiledApp, then in the makefile in that directory I would add the following code? Thx!

MODULE_NAME := heat_conduction

**Reply by GiudGiud** [reply]
So people s app generally dont live in a module, they have a different folder.
But if you want to stay like this, then modify the modules.mk file in framework/modules.

But actually, tensor mechanics is already a dependence of phase_field. So there should be no need to move files

**Comment by jsevic** [comment]
ok is there a location in the moose website that might help better understand this process, may as well try to get over as many of my bad habits too since i got this new machine set up to continue my work thx!

**Reply by GiudGiud** [reply]
https://mooseframework.inl.gov/getting_started/new_users.html#create-an-app is the one i thin

**Comment by jsevic** [comment]
Ah yes ok I remember this now, and I remember not doing it, so essentially I create a clone of moose, except myNewApp only compiles with the modules I selected in the MakeFile, so if I want the phase_field and heat_conduction modules in my new app then I would select these. Does this process also bring along any dependencies that either of these two modules may require? Once I compile myNewApp then all of the kernels in these two modules are available to me?

Very nice I should have learned earlier thx!

**Reply by GiudGiud** [reply]
yes it brings both these module s dependencies. And all the kernels & objects in those modules can be called from the app.

**Comment by jsevic** [comment]
That is pretty sweet - I added those modules to my new app tiger and I can run both my heat and PF tests. Hopefully this gets me going and I certainly learned a lot more about moose and the pathological way I was running my sums thanks so much.

**Comment by jsevic** [comment]
Another question on the new kernel Heat Conduction, in a previous MOOSE app I made a custom kernel for charge conservation that used as a proxy for electrical conductivity c, the P-F concentration variable. I explicitly coupled this in the .c code, and this worked.

Now when I use Heat Conduction and pass c using “diffusion coefficient” I get the following error…

Material property ‘c’ requested by…

Does this mean that these new kernels don’t allow a state variable to be coupled?

Thank you!

**Reply by GiudGiud** [reply]
Hello

Did you modify those kernels?
https://mooseframework.inl.gov/source/kernels/HeatConduction.html , the one in MOOSE, expects a `diffusion_coefficient` material property, not a variable. 

You could use a ParsedMaterial
https://mooseframework.inl.gov/source/materials/ParsedMaterial.html
to convert the variable into a material property.

Guillaume

**Reply by GiudGiud** [reply]
I cant remember, are you working with AD?

**Comment by jsevic** [comment]
I used JouleHeatingSource and HeatConduction, and in my earlier app from 2018, I made a custom kernel that started with the MOOSE diffusion operator and I added a coupled variable to it for a new kernel. So it looks like these two kernels JouleHeatingSource and HeatConduction are expecting values from material blocks, and for that I settled on SemiconductorLinearConductivity since I was able to fit A and B to my material system.

If I understand what you are saying, for example I am using DerivativeParsedMaterial and in this material I am passing as arguments state variables from the kernels using the arg command. In a similar way, I could use ParsedMaterial if I don’t need the Jacobian and then here also I could pass in arg state variables and then do something on them and then f_name = NewParsedVariable could be placed in the diffusion_coefficient input of HeartConduction for example?


Thx!

John

**Reply by GiudGiud** [reply]
Yes you're right. 
If you dont use PJFNK (if you use Newton for example), you'll get a better Jacobian by specifying the parameter `diffusion_coefficient_dT` to HeatConduction.
To create this derivative, you can create your own material or you can use the DerivativeParsedMaterial https://mooseframework.inl.gov/source/materials/DerivativeParsedMaterial.html
The documentation shows how to create the derivative
